### PR TITLE
LOG4J2-3280 - Add support for microsecond precision in RFC5424Layout

### PR DESCRIPTION
**Pull request by atulpendse**
This pull request is to imlement RFE LOG4J2-3280
The change 
  - Adds new attribute timestampPrecision to Rfc5424Layout and SyslogAppender.
  - Replaces existing logic to append timestamp and uses two new fixed date formats to format timestamp.
  - Added new unit tests
  - Verified that 'mvn clean install' and 'mvn verify' are successful.

### CODE REVIEW

**Comment by atulpendse** [comment]
I believe the failure in PropertiesConfigurationTest is unrelated to this change, and being fixed with #685 
Once that is fixed, how do I get this PR to build and run CI tests again?

**Comment by garydgregory** [comment]
Hi @atulpendse
You will need to rebase this PR to pick up the latest changes.

**Comment by atulpendse** [comment]
> Hi @atulpendse
> You will need to rebase this PR to pick up the latest changes.

Thanks for the info. I have updated PR with the latest from upstream. Latest commit seems to have passed all the tests.

**Comment by vy** [comment]
Fantastic work @atulpendse! Thanks so much!

Mind adding `SyslogAppender` tests and a `changes.xml` entry, please?

I am also not sure if a `timestampPrecision` argument of type `String` is the right way to go. Thinking out loudly:
- Why don't we simply make the instant formatting pattern configurable?
- Doesn't the spec enforce a certain precision?
- If not, why don't we simply (automatically?) pick the most precise representation possible?

**Comment by garydgregory** [comment]
I would NOT update changes.xml, this makes cherry-picking the commit to master impossible.

**Comment by vy** [comment]
IMHO, cherry-picking to `master` is almost already impossible. Nevertheless, I will rephrase my comment to exclude the `changes.xml` part.

**Comment by atulpendse** [comment]
@vy 
Thanks for taking time to review the PR.
Here are my thoughts about questions you raised.

> I am also not sure if a `timestampPrecision` argument of type `String` is the right way to go. Thinking out loudly:
> 
> * Why don't we simply make the instant formatting pattern configurable?

With making pattern configuration, there is risk of the message not being compliant to RFC5424 spec (if someone updates format that is not as per RFC5424)

> * Doesn't the spec enforce a certain precision?

The spec does not enforce 'a particular' precision, rather it supports three precisions

1. Without any sub-second value
2. Milliseconds
3. Microseconds


> * If not, why don't we simply (automatically?) pick the most precise representation possible?

We could, that would work for my usecase. But if someone is expecting precision to be millisecond (like my app was expecting it to be microsecond), it would break their app.

I will add a test for SyslogAppender.

**Comment by atulpendse** [comment]
Commit fbccefc updates existing test for SyslogAppender

**Comment by atulpendse** [comment]
@vy @garydgregory anything else that I should do to make progress on this PR?

File: log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java [hunk 1]
```diff
@@ -211,6 +214,10 @@ public String getExceptionPattern() {
             return loggerFields;
         }
 
+		public String getTimestampPrecision() {
```
Comment (garydgregory) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java:217]: This indentation looks messed up. Spaces, no tabs, could be the issue?

File: log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java [hunk 2]
```diff
@@ -300,6 +307,11 @@ public B setLoggerFields(final LoggerFields[] loggerFields) {
             this.loggerFields = loggerFields;
             return asBuilder();
         }
+
+		public B setTimestampPrecision(final String timestampPrecision) {
```
Comment (garydgregory) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java:311]: This indentation looks messed up. Spaces, no tabs, could be the issue?

File: log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java [hunk 3]
```diff
@@ -104,6 +104,9 @@
         @PluginElement("LoggerFields")
         private LoggerFields[] loggerFields;
 
+        @PluginBuilderAttribute( "timestampPrecision" )
+        private String timestampPrecision;
```
Comment (garydgregory) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java:108]: Shouldn't we use an enum here?

File: log4j-core/src/main/java/org/apache/logging/log4j/core/layout/Rfc5424Layout.java [hunk 4]
```diff
@@ -270,7 +278,7 @@ private static PatternParser createPatternParser(final Configuration config,
     public String toSerializable(final LogEvent event) {
         final StringBuilder buf = getStringBuilder();
         appendPriority(buf, event.getLevel());
-        appendTimestamp(buf, event.getTimeMillis());
+		appendTimestamp(buf, event.getInstant());
```
Comment (garydgregory) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/layout/Rfc5424Layout.java:281]: No tabs.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java [hunk 5]
```diff
@@ -425,6 +440,7 @@ protected SyslogAppender(final String name, final Layout<? extends Serializable>
                 .setCharsetName(charset)
                 .setExceptionPattern(exceptionPattern)
                 .setLoggerFields(loggerFields)
+                .setTimestampPrecision( timestampPrecision )
```
Comment (garydgregory) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java:443]: No spaces around an argument, please follow the style of the file you're in!
`.setTimestampPrecision( timestampPrecision )` -> `.setTimestampPrecision(timestampPrecision)`

**Review by garydgregory — CHANGES_REQUESTED** [review]
I only took a cursory look this AM. Please use the style of the files you are editing: Use spaces, no tabs. The bigger question is whether we should use an enum for the microsecond vs millisecond format, @jvz any thoughts?

File: log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java (hunk 3 above)
Comment (jvz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java:108]: I agree. An enum would make sense.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java (hunk 1 above)
Comment (atulpendse) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java:217]: Yeah.. issue due to tabs. Was due to problem with my eclipse formatter setting. I have fixed it with latest commit.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java (hunk 2 above)
Comment (atulpendse) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java:311]: I have fixed it with latest commit.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java (hunk 5 above)
Comment (atulpendse) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java:443]: I have fixed it with latest commit.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/layout/Rfc5424Layout.java (hunk 4 above)
Comment (atulpendse) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/layout/Rfc5424Layout.java:281]: I have fixed it with latest commit.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java (hunk 3 above)
Comment (atulpendse) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java:108]: I have converted it into an enum with latest commit.

**Comment by atulpendse** [comment]
> I only took a cursory look this AM. Please use the style of the files you are editing: Use spaces, no tabs. The bigger question is whether we should use an enum for the microsecond vs millisecond format, @jvz any thoughts?

Thanks for taking time to review. I have fixed the concerns with latest commit. Now timestampPrecision is an enum.

**Review by garydgregory — COMMENTED** [review]
Looks good. I approved the PR for builds here, let's see what happens, especially with the GC tests as I am not sure if this class falls into the GC free epic we had.

**Comment by atulpendse** [comment]
Looks like a test is failing on MacOS. Failure doesn't look related to my changes though. I don't have access to Mac to find out why exactly it's failing.

**Comment by atulpendse** [comment]
@garydgregory, please let me know how I can help to get the PR going.
I rebased the PR with latest code and submitted back with commit b425497

**Comment by atulpendse** [comment]
Just FYI.. I was able to build it successfully on my Mac. Last time @garydgregory tried to submit build with this PR, it failed on Mac, not sure why.

File: log4j-core/revapi.json [hunk 6]
```diff
@@ -128,6 +128,25 @@
         "annotation": "@org.apache.logging.log4j.core.config.plugins.PluginFactory",
         "justification": "LOG4J2-3317: Fix RoutingAppender backcompat while improving lookup security"
       },
+      {
+        "code": "java.method.numberOfParametersChanged",
```
Comment (garydgregory) [inline log4j-core/revapi.json:132]: You should deprecate the old method, not remove it.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/layout/Rfc5424Layout.java [hunk 7]
```diff
@@ -627,7 +608,8 @@ public static Rfc5424Layout createLayout(
             // RFC 5425
             @PluginAttribute(value = "useTlsMessageFormat") final boolean useTlsMessageFormat,
             @PluginElement("LoggerFields") final LoggerFields[] loggerFields,
-            @PluginConfiguration final Configuration config) {
+            @PluginConfiguration final Configuration config,
+            @PluginAttribute(value = "timestampPrecision") final TimestampPrecision timestampPrecision) {
```
Comment (garydgregory) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/layout/Rfc5424Layout.java:612]: Use a builder instead of breaking BC.

**Comment by garydgregory** [comment]
Hello @atulpendse 
In general, you want to avoid breaking binary compatibility,  so that means not editing public and protected type names and method signatures. This is critical in the log4j-api module and less so in others, but, in this case it seems easy to do. Also, we have been switching from factory methods to builders as we go from release to release in order to avoid keeping on deprecating factory methods. So edit or add builders, don't edit factory methods is the idea. TY!

File: log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java [hunk 8]
```diff
@@ -393,7 +408,8 @@ protected SyslogAppender(final String name, final Layout<? extends Serializable>
             final Charset charset,
             final String exceptionPattern,
             final LoggerFields[] loggerFields,
-            final boolean advertise) {
+            final boolean advertise,
+            final TimestampPrecision timestampPrecision) {
         // @formatter:on
```
[the remark this answers is not in the document]
Comment (atulpendse) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/appender/SyslogAppender.java:413]: I agree, but Rfc5424Layout does not have a builder which caller can use. It only supports creating instances via constructor.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/layout/Rfc5424Layout.java [hunk 9]
```diff
@@ -92,11 +97,12 @@
      */
     public static final String DEFAULT_MDCID = "mdc";
 
+    private static final FixedDateFormat FIXED_DATE_FORMAT_MILLI_PRECISION = FixedDateFormat
+            .create(FixedFormat.ISO8601_OFFSET_DATE_TIME_HHCMM_PERIOD);
+    private static final FixedDateFormat FIXED_DATE_FORMAT_MICRO_PRECISION = FixedDateFormat
+            .create(FixedFormat.ISO8601_OFFSET_DATE_TIME_HHCMM_PERIOD_MICRO);
+
```
Comment (ppkarwasz) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/layout/Rfc5424Layout.java:104]: I would use a (non static) `org.apache.logging.log4j.core.pattern.DatePatternConverter` to keep the layout garbage free.

**Comment by rgoers** [comment]
@vy Why was this closed? To be honest, I am not sure why this hasn't been merged.

**Comment by ppkarwasz** [comment]
It's a side effect of deleting the `release-2.x` branch: Github deletes the related PRs.

@atulpendse: can submit again against the `2.x` branch?

**Pull request closed without merging**